docs/specification.md

```diff
@@ -465,57 +506,115 @@ data Parameter proposalMetadata
+
+```haskell
+data Parameter proposalMetadata
+  = SetVotingPeriod Natural
+```
+
+Parameter (in Michelson):
+```
+nat
+```
+
+- Update how long the voting period lasts for new proposals
+- Fails with `NOT_ADMINISTRATOR` if the sender is not the administrator.
+
+### **set_quorum_threshold**
+
+```haskell
+data Parameter proposalMetadata
+  = SetQuorumThreshold Natural
+```
+
+Parameter (in Michelson):
+```
+nat
+```
+
+- Update the amount of votes which is required for a successful proposal for new proposals
+- Fails with `NOT_ADMINISTRATOR` if the sender is not the administrator.
 
 ### **vote**
 
 ```haskell
-type Parameter proposalMetadata
-  = Vote VoteParam
+type ProposalId = Natural
 
 type VoteType = Upvote | Downvote
 
 data VoteParam = VoteParam
-  { vpProposalId :: ProposalId
-  , vpVoteType :: VoteType
+  { proposal_id :: ProposalId
+  , vote_type :: VoteType
   }
+
+data Parameter proposalMetadata
+  = Vote VoteParam
 ```
 
 Parameter (in Michelson):
+
 ```
-TODO
+(pair %vote
+  (nat %proposal_id)
+  (bool %vote_type)
+)
 ```
 
-- Sender MUST have frozen token (`token_id = 1`) balance not associated with others proposals,
-  equal to or more than the amount specified by the target proposal.
+- The `Natural` value (let's say `n`) denotes the amount of tokens required by the proposal to vote.
+- Sender MUST have enough unfrozen tokens (i. e. `≥ n`).
+- Fails with `VOTING_INSUFFICIENT_BALANCE` if the unfrozen token balance of the SENDER
+  is less than `n`.
+- Fails with `PROPOSAL_NOT_EXIST` if the proposal id is not associated with any ongoing proposals.
+- The sender's balance in frozen tokens is increased by `n` and in unfrozen tokens is decreased by `n`.
+
 
-- After a successful vote, the SENDER's frozen tokens amount specified by the proposal are not associated
-  with the proposal and will not be able to use to vote for other contract.
+### **flush**
 
-- Fails with `PROPOSAL_NOT_EXIST` if the proposal id is
+```haskell
+data Parameter proposalMetadata
+  = Flush
+```
 
-- After the target proposal period ends:
-  - The sender frozen token associated with the proposal will be deducted from
-    his frozen token balance.
-  - The sender unfrozen token balance will be transferred from `freeze_my_token` address
-    with the amount equally to frozen tokens associated with the contract.
-    This transfer must follow FA2 requirement's `transfer`.
+Parameter (in Michelson):
+```
+unit
+```
 
-not associated with any on going proposals.
+- Finish voting process on all proposals for which voting period is over.
+- Frozen tokens from voters and proposal submitter associated with those proposals are returned
+  in form of unfrozen tokens:
+  - If proposal got reject: the return amount are slashed base on `cRejectedProposalReturnValue`.
+  - If proposal got accepted: the return amount are returned in total.
 
-### **flush**
+## Migrating functions
+
+### **migrate**
+
+```haskell
+data Parameter proposalMetadata
+  = Migrate Address
```

Comment:
Actually I intentionally removed Haskell definitions for entrypoints with trivial types like this one. Michelson representation should be treated as source of truth, because all tools can use it. Haskell definitions are provided only for readability. When there is a type with multiple fields or constructors, they are useful, but for atomic types like `address` it looks redundant.

That being said, if you really want to have Haskell definitions for all types (e. g. for consistency), I am kinda ok with that. So I won't remove them anymore, I am leaving this decision to you.

Reply:
I didn't know you remove it 😅. I'm okay with both, and like you said, Ivan, I only put it for consistency.